lib/tdf3/src/tdf.ts

```diff
@@ -941,60 +941,64 @@ export class TDF extends EventEmitter {
-            that.manifest?.encryptionInformation.integrityInformation.segmentHashAlg;
-          const segmentHashStr = await that.getSignature(
-            reconstructedKeyBinary,
-            Binary.fromBuffer(encryptedChunk),
-            segmentIntegrityAlgorithmType || integrityAlgorithmType
-          );
+    // const that = this;
 
-          if (segment.hash !== base64.encode(segmentHashStr)) {
-            throw new ManifestIntegrityError('Failed integrity check on segment hash');
-          }
+    const underlyingSource = {
+      pull : async (controller: ReadableStreamDefaultController) => {
 
-          let decryptedSegment;
+        if (segments.length === 0) {
+          controller.close();
+          return;
+        }
 
-          try {
-            decryptedSegment = await cipher.decrypt(encryptedChunk, reconstructedKeyBinary);
-          } catch (e) {
-            throw new TdfDecryptError(
-              'Error decrypting payload. This suggests the key used to decrypt the payload is not correct.',
-              e
-            );
-          }
+        const segment = segments.shift();
+        if (!segment) {
+          throw new Error('Shifted past end of segments array');
+        }
+        const encryptedSegmentSize = segment?.encryptedSegmentSize || encryptedSegmentSizeDefault;
+        const encryptedChunk = await zipReader.getPayloadSegment(
+          centralDirectory,
+          '0.payload',
+          encryptedOffset,
+          encryptedSegmentSize
+        );
+        encryptedOffset += encryptedSegmentSize;
+
+        // use the segment alg type if provided, otherwise use the root sig alg
+        const segmentIntegrityAlgorithmType =
+          this.manifest?.encryptionInformation.integrityInformation.segmentHashAlg;
+        const segmentHashStr = await this.getSignature(
+          reconstructedKeyBinary,
+          Binary.fromBuffer(encryptedChunk),
+          segmentIntegrityAlgorithmType || integrityAlgorithmType
+        );
 
-          controller.enqueue(decryptedSegment.payload.asBuffer());
+        if (segment?.hash !== base64.encode(segmentHashStr)) {
+          throw new ManifestIntegrityError('Failed integrity check on segment hash');
         }
 
-        if (segments.length === 0) {
-          controller.close();
+        let decryptedSegment;
+
+        try {
+          decryptedSegment = await cipher.decrypt(encryptedChunk, reconstructedKeyBinary);
+        } catch (e) {
+          throw new TdfDecryptError(
+            'Error decrypting payload. This suggests the key used to decrypt the payload is not correct.',
+            e
+          );
         }
+
+        controller.enqueue(decryptedSegment.payload.asBuffer());
       },
-    });
+    };
+
+    const outputStream = makeStream(underlyingSource);
+
+    if (rcaParams && rcaParams.wu) {
+      let res = await axios.head(rcaParams.wu);
+
+      outputStream.fileSize = parseInt(res?.headers?.['content-length'] as string, 10) || undefined;
```

Comment:
You may want to log some details here if this doesn't work, as we probably want this to be an error or warning for us if it fails, since it indicates a configuration problem with the storage server